src/selectors/createSelector.ts

```diff
@@ -0,0 +1,91 @@
+import {Selector} from 'testcafe-without-typecheck';
+
+import {DESCRIPTION_KEY} from '../constants/internal';
+
+import type {Fn, RawSelector, SelectorCustomMethods, Values} from '../types/internal';
+
+/**
+ * Proxy handler for wrapping all selector properties.
+ */
+const createGet = <CustomMethods extends SelectorCustomMethods = {}>(
+  customMethods?: CustomMethods,
+): Required<ProxyHandler<RawSelector<CustomMethods>>>['get'] => {
+  const get: Required<ProxyHandler<RawSelector<CustomMethods>>>['get'] = (
+    target,
+    property,
+    receiver,
+  ) => {
+    const customMethod =
+      customMethods && typeof property === 'string' ? customMethods[property] : undefined;
+
+    let result = (
+      customMethod ? customMethod.bind(target) : Reflect.get(target, property, receiver)
+    ) as Values<RawSelector<CustomMethods>> & {
+      [DESCRIPTION_KEY]?: string;
+    };
+
+    if (typeof property === 'symbol') {
+      return result;
+    }
+
+    if ((typeof result !== 'object' || result === null) && typeof result !== 'function') {
+      return result;
+    }
+
+    if (typeof result === 'function') {
+      const originalFunction = result as Fn;
+
+      result = // eslint-disable-next-line no-restricted-syntax
+        function selectorMethodWrapper(this: RawSelector, ...args: never[]) {
+          const callResult = originalFunction.apply(this, args);
+
+          if (
+            (typeof callResult !== 'object' || callResult === null) &&
+            typeof callResult !== 'function'
+          ) {
+            return callResult;
+          }
+
+          const callLocator = `${result[DESCRIPTION_KEY]}(${args.join(', ')})`;
+
+          (callResult as typeof result)[DESCRIPTION_KEY] = callLocator;
+
+          // callResult is Selector
+          if (Object.prototype.hasOwnProperty.call(callResult, 'getBoundingClientRectProperty')) {
+            return new Proxy(callResult, {get});
+          }
+
+          return callResult;
+        } as typeof result;
+    }
+
+    const locator = target[DESCRIPTION_KEY] || '';
+
+    result[DESCRIPTION_KEY] = `${locator}.${property}`;
+
+    return result;
+  };
+
+  return get;
+};
+
+export type CreateSelector<CustomMethods extends SelectorCustomMethods = {}> = (
```

Comment:
`suggestion`: I would also move this type to `src/types/selectors`, in accordance with the [dependency inversion principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle) — all reused types are in `src/types`, forming abstractions separated from implementation details. Implementations import and use these types. When refactoring the implementation (without API changes), the `src/types` directory is not affected.

Reply:
i thought dependency inversion principle not about that, but i moved this type to /types/selectors to be consistent